src/invert/laplace/impls/iterative_parallel_tri/iterative_parallel_tri.cxx

```diff
@@ -0,0 +1,1597 @@
+  // proc: |       p-1       |          p          |       p+1
+  //    x: |            xs-1 | xs      ...      xe | xe+1
+  // xloc: |xloc[0] ...      | xloc[1] ... xloc[2] | xloc[3]
+  //
+  // In this method, we write the original tridiagonal system as a reduced
+  // tridiagonal system whose grid points are the first interior point on each
+  // processor (xs), plus the last interior point on the final processor. Each
+  // processor solves for its local points. Quantities are stored in length 4
+  // arrays where index 1 corresponds to the value at a processor's xs on the
+  // original grid. Indices 0 and 3 correspond to the lower and upper
+  // neighbours' values at their xs respectively, except on physical
+  // boundaries where these are the boundary points. Index 2 is used on the
+  // final processor only to track its value at the last grid point xe. This
+  // means we often have special cases of the equations for final processor.
+  //
+  auto xloc = Matrix<dcomplex>(4,nmode); // Values of x at the processor interface
+
+  BoutReal kwaveFactor = 2.0 * PI / coords->zlength();
+
+  // Should we store coefficients?
+  store_coefficients = not (inner_boundary_flags & INVERT_AC_GRAD);
+  store_coefficients = store_coefficients && not (outer_boundary_flags & INVERT_AC_GRAD);
+  store_coefficients = store_coefficients && not (inner_boundary_flags & INVERT_SET);
+  store_coefficients = store_coefficients && not (outer_boundary_flags & INVERT_SET);
+
+  // Setting the width of the boundary.
+  // NOTE: The default is a width of 2 guard cells
+  int inbndry = localmesh->xstart, outbndry=localmesh->xstart;
+
+  // If the flags to assign that only one guard cell should be used is set
+  if((global_flags & INVERT_BOTH_BNDRY_ONE) || (localmesh->xstart < 2))  {
+    inbndry = outbndry = 1;
+  }
+  if (inner_boundary_flags & INVERT_BNDRY_ONE)
+    inbndry = 1;
+  if (outer_boundary_flags & INVERT_BNDRY_ONE)
+    outbndry = 1;
+
+  /* Allocation for
+  * bk   = The fourier transformed of b, where b is one of the inputs in
+  *        LaplaceIPT::solve()
+  * bk1d = The 1d array of bk
+  * xk   = The fourier transformed of x, where x the output of
+  *        LaplaceIPT::solve()
+  * xk1d = The 1d array of xk
+  */
+  auto bk = Matrix<dcomplex>(ncx, ncz / 2 + 1);
+  auto bk1d = Array<dcomplex>(ncx);
+  auto xk = Matrix<dcomplex>(ncx, ncz / 2 + 1);
+  auto xk1d = Matrix<dcomplex>(ncz/2+1,ncx);
+
+  // Error interpolated onto a finer grid
+  auto fine_error = Matrix<dcomplex>(4,ncz/2+1);
+
+  ///SCOREP_USER_REGION_END(initvars);
+  ///SCOREP_USER_REGION_DEFINE(initloop);
+  ///SCOREP_USER_REGION_BEGIN(initloop, "init xk loop",SCOREP_USER_REGION_TYPE_COMMON);
+
+  // Initialise xk to 0 as we only visit 0<= kz <= maxmode in solve
+  for (int ix = 0; ix < ncx; ix++) {
+    // TODO This looks wrong: should it be kz<ncz?
+    // This came from serial tri
+    for (int kz = maxmode + 1; kz < ncz / 2 + 1; kz++) {
+      xk(ix, kz) = 0.0;
+    }
+  }
```

Comment:
This works, because the correct part of `xk` gets overwritten by the solution at the end.  It's only part of the array that's set to zero in the original loop though.  Also, as the comment says, it looks funny because `maxmode = ncz / 2`, so I think this might be a typo for `for (int kz = maxmode + 1; kz < ncz; kz++) {`.

Reply:
Ah, I see! I completely missed the start index for `kz`. Still, probably doesn't hurt to initialise it all to `0.0` anyway?

Reply:
At the moment it's doing nothing, so it might be better just to delete it.